accepted/00065-salt-minions-update-cycle.md

```diff
@@ -0,0 +1,127 @@
+
+# Motivation
+[motivation]: #motivation
+
+Due to Salt's nature, it comes with the modules that are persistently installed on the Salt Minions. Given Salt Minion outdated due to variety of reasons (e.g. being a part of OS's maintenance schedule), this leads to inconsistent infrastructure, where the same modules might have different capabilitles.
+
+This RFC has the following goals to resolve:
+
+- Ensure Python environment with a specific version on all clients.
+- Ensure identical version of Salt Minions with the Salt Master.
+- Independence from client OS's maintenance cycle.
+
+# Detailed design
+[design]: #detailed-design
+
+## Approach
+
+**Independent Maintenance Cycle**
+
+In order to efficiently update Salt Minion on any version of operating system, the update cycle of a Salt Minion needs to be independent from the maintenance update of the very operating system itself.
+
+Rationale: Allow components update as soon as it is needed for the software component, without waiting for the general maintenance update.
+
+**Isolation**
+
+Since Salt Minion is a part of configuration management system, it is also a good to have it "static", in terms of zero dependencies from the operating system it currently runs.
+
+Rationale: Allow running latest modern software on much older environments without any impact to the managed operating system.
+
+**In-place Bootstrapping**
+
+Each Salt Minion is not packaged anymore in OS-specific package manager (RPM, Deb etc), but is bootstrapped via standard Python mechanism (Package Installer for Python, PIP).
+
+Rationale: Track all the patches in only one place. As the code is isolated (see above), avoid unnecessary extra-packaging for isolated version of the Salt Minion.
+
+## Principle
+
+The principle is to run own minimal [PyPi](https://pypi.org/project/pypiserver/) local server on Uyuni cluster. In order to install a Salt Minion, a managed system should use PIP againstlocal PyPi server, as shown on Figure 1 below:
+
+![Layout](images/00065-layout.png)
+
+_Figure 1_
+
+Process update or installation flow as follows:
+1. One RPM general package of Salt Minion **(1)** is provided for all possible platforms and architectures and is installed on PyPi server. However, Salt Minion is not _installed_ per se in the sense as for end-user, but as a PyPi package into the PyPi repository location. This way patches are still tracked at RPM `.spec` level.
+2. General statically compiled Python vanilla environment is provided as an RPM package from Uyuni repository and is allowed to be installed **(2)** on a Managed System, e.g. in `/opt/python3.8` or whatever version is needed. That way it is also allowed to have many Python versions as well as its modifications, if ever needed.
+3. Once virtual environment (not shown on the _Figure 1_) is created for the the separate Python interpreter, PIP is bootstrapping the entire Salt Minion against Uyuni-served PyPi server with all the required Salt Minion dependencies.
```

Comment:
I was not thinking about about Salt SSH. If you had read carefully, I was already mentioned watchdog function in the section of Fail-proof installation, as a part of the virtual env manager. The Virtual Env manager is not much of a rocket science, but just a simple program that would switch links between shell scripts that starts environments. The same program (implemented in whatever way) would also happily create virtual environments.

The OBS venvjail is an interesting topic. However, it is operating in different paradigm? I was not thinking applying PTF on each _client Minion_ separately, but once on the _Uyuni Server's PyPi server_, so then Salt Minion will just update itself from it, applying the patch.

Reply:
The whole idea is to detach Salt Minion from the system completely, so direct removal of it by `rm -rf /opt/salt` won't hurt anything of the operating system itself. I don't mind shipping static Python as just a tar archive, served somewhere on Uyuni server over HTTP.

Reply:
> I was not thinking about about Salt SSH. If you had read carefully, I was already mentioned watchdog function in the section of Fail-proof installation,

I did. My question remains. Who will create the venv?

You say this watchdog (even watchgod are usually only for monitoring and triggering). Ok, the watchdog will create the venv.

So the client will install the watchdog in the node that want to register. I guess that it found the there is no venv, somehow will find the pip registry in the DNS, somehow will know where the static Python is living, will also know the order of package installation (or where is living the requirements.txt for the first venv), etc, etc.

I think that you need to dig into those details, as will help how big / small are the real work.

Also managing the live-cycle of a venv is not trivial. I know that because in Ardana we had exactly this approach that you are talking about, but using Ansible (that resolve the bootstrap question). Until we moved to OBS + venvjail, the machinery required to bootstrap the initial venv for each service was insane. Full of details and exceptions, put in place based on real-world experience.

> as a part of the virtual env manager — something that I don't find yet a time to talk about.

So, please, dig here into the rabbit hole.

> The OBS venvjail is an interesting topic. However, it is operating in different paradigm? I was not thinking applying PTF on each _client Minion_ separately, but once on the _Uyuni Server's PyPi server_, so then Salt Minion will just update itself from it, applying the patch.

Sure, but when a single minion is failing, the first thing that you need to know is what is inside the venv. The exact version that is running, the changelogs, etc.

Do you expect that all the venvs will be flawless updated, atomically, ever? Let me put some examples:

* The fact that the pypi local registry have PyYAML version 2.4.0-45, that have this important patch released by Pablo two weeks ago, do not imply that is available in the minion that is failing.

* L3 needs to send a PTF, but is a test one, that will be target to one single node. How can be done without replacing the wheel in the local registry? How can be targeted to a single node?

* How can you trace the wheel to the exact version in IBS / OBS of the package, to check if a CVE was already fixed or not?

> Generally, speaking, it should be possible to just flush the entire Salt Minion away and re-bootstrap it, if it is needed (preserving keys, config and machine ID parts).

I agree. But this is not how the daily support works. You need to make it easy to find the root cause, and track the problem and the fix. All this can be fixed and provided, but is not a small amount of work.

Reply:
> when a single minion is failing, the first thing that you need to know is what is inside the venv. The exact version that is running, the changelogs, etc.

This is no different how SaltSSH is failing.

Reply:
The whole approach is boiling down to make Salt Minion volatile in the same fashion as SaltSSH is already doing anyway. So the whole story of the patching applies the same path as it is applies to patch SaltSSH `thin.tgz`.  However, it seems unclear from the RFC at the moment.  I will update the RFC and re-request review.